dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+        related to the joint and marginal distributions.
+
+        Parameters
+        ----------
+        nbins : int
+            the number of bins of the joint and marginal probability density
+            functions (the actual number of bins of the joint PDF is nbins**2)
+
+        References
+        ----------
+        [Mattes03] Mattes, D., Haynor, D. R., Vesselle, H., Lewellen, T. K.,
+                   & Eubank, W. PET-CT image registration in the chest using
+                   free-form deformations. IEEE Transactions on Medical Imaging,
+                   22(1), 120-8, 2003.
+        """
+        self.nbins = nbins
+        # Since the kernel used to compute the Parzen histogram covers more than
+        # one bin, we need to add extra bins to both sides of the histogram
+        # to account for the contributions of the minimum and maximum
+        # intensities. Padding is the number of extra bins used at each side
+        # of the histogram (a total of [2 * padding] extra bins). Since the
+        # support of the cubic spline is 5 bins (the center plus 2 bins at each
+        # side) we need a padding of 2, in the case of cubic splines.
+        self.padding = 2
+
+    def setup(self, static, moving, smask=None, mmask=None):
+        r""" Compute histogram settings to store the PDF of input images
+
+        Parameters
+        ----------
+        static : array
+            static image
+        moving : array
+            moving image
+        smask : array
+            mask of static object being registered (a binary array with 1's
+            inside the object of interest and 0's along the background).
+            If None, the behaviour is equivalent to smask=ones_like(static)
+        mmask : array
+            mask of moving object being registered (a binary array with 1's
+            inside the object of interest and 0's along the background).
+            If None, the behaviour is equivalent to mmask=ones_like(static)
+        """
+
+        if smask is None:
+            smask = np.ones_like(static)
+        if mmask is None:
+            mmask = np.ones_like(moving)
+
+        self.smin = np.min(static[smask!=0])
+        self.smax = np.max(static[smask!=0])
+        self.mmin = np.min(moving[mmask!=0])
+        self.mmax = np.max(moving[mmask!=0])
+
+        self.sdelta = (self.smax - self.smin)/(self.nbins - 2 * self.padding)
+        self.mdelta = (self.mmax - self.mmin)/(self.nbins - 2 * self.padding)
+        self.smin = self.smin/self.sdelta - self.padding
+        self.mmin = self.mmin/self.mdelta - self.padding
+
+        self.joint_grad = None
+        self.metric_grad = None
+        self.metric_val = 0
+        self.joint = np.ndarray(shape = (self.nbins, self.nbins))
+        self.smarginal = np.ndarray(shape = (self.nbins,), dtype = np.float64)
+        self.mmarginal = np.ndarray(shape = (self.nbins,), dtype = np.float64)
+
+
+    def bin_normalize_static(self, x):
```

Comment:
True, it's not being called from Cython, I just added the method for convenience in case the user needs it. It is being tested, though.